Fix EBADF error using broadcast

heisters opened the pull request:
Broadcast was unusable because setBroadcast has to be set after the socket is opened or it results in an EBADF error.

I also added a broadcast option, since it was unclear that the multicast option would turn on broadcast. Lastly, I removed the custom defaults to local address and local port because they are inconsistent with the expected behavior of dgram, which defaults to listening on all interfaces (0.0.0.0) instead of just localhost.

A working broadcast example now looks like this:

````
var osc = require('osc');

var oscPort = new osc.UDPPort({
  localPort: 5000,
  remoteAddress: '192.168.1.255',
  remotePort: 5000,
  broadcast: true
});
oscPort.on("open", function(socket) {
  setInterval( function() {
    oscPort.send({
      address: "/test",
      args: []
    });
  }, 1000 );
});

oscPort.open();
````

And to test reception:

````
var port = process.argv[2]
  , address = '0.0.0.0'
  , dgram = require('dgram')
;

if ( port === undefined ) {
  console.log( "Please specify a port." );
  console.log( process.argv[1] + " [PORT]" );
  process.exit(1);
}

var server = dgram.createSocket('udp4')
server.on("listening", function() {
  var a = server.address();
  console.log( "Listening on " + a.address + ":" + a.port );
});

server.on("message", function(message, remote) {
  console.log( remote.address + ":" + remote.port + " - " + message );
});

server.bind(port, address);
````

colinbdclark commented:
Hey @heisters, this is awesome! Thanks so much!

Can you adapt your example above into a unit test case so that we can easily verify this new feature and ensure that it doesn't regress in the future? There are a variety of other tests for osc,js' Node.js UDP support located in the [node-transport-tests.js](https://github.com/colinbdclark/osc.js/blob/master/tests/node-transport-tests.js) file.

Also, I'm not sure if it makes sense to remove the defaults for localPort and localAddress, though it does look like there is currently some inconsistency in how this is handled between the multicast and unicast modes. My goal is to make osc.js as simple and safe as possible for non-expert programmers. In the typical use case of using osc.js to relay OSC messages between a web browser and a locally-installed music application such as Max, SuperCollider, or Pd, it is safer and more secure to default to 127.0.0.1 than to all addresses on the system (including Internet-visible addresses). Does this make sense?

Thanks again!

heisters commented:
Ok, that reasoning makes sense. I think I was responding to the fact that if you didn't set `localAddress` and you did set `multicast`, you would end up with a different default than if you didn't set `multicast`. Passing localAddress to bind regardless of multicast, broadcast, or otherwise, fixes that. 

To keep things consistent, I added the same defaults to remote port/address so that the default is a client/server pair that talks to itself.

I added tests for multicast and broadcast. I question the multicast flag entirely, since it won't *really* work without other multicast configuration like addMembership. Nevertheless, I left it in but added a broadcast flag that works more like I would expect.

The broadcast test just fails by hanging since the message is never received if setBroadcast is not called. I'm not sure how to make the test timeout and fail outright. Also, I think a remoteAddress of 255.255.255.255 might be an error on some platforms or configurations, but there's no other simple way of sending broadcast packets without manual configuration.

colinbdclark commented:
Awesome, thanks so much. I'll take a look at the test and see if I can get it working. I really appreciate your help, this is great.

heisters commented:
Oh, the test passes! It's just not a clean red failure when the code that
makes it work is removed. Which has more to do with the test suite setup
than the test itself, since async tests should probably time out and fail.
On Tue, Aug 4, 2015 at 7:59 AM Colin Clark <notifications@github.com> wrote:

> Awesome, thanks so much. I'll take a look at the test and see if I can get
> it working. I really appreciate your help, this is great.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/colinbdclark/osc.js/pull/23#issuecomment-127641773>.
>
